### PR TITLE
Add prometheus ingress

### PR DESCRIPTION
**Pull request by omegas27**
Towards https://github.com/giantswarm/giantswarm/issues/13366
This still needs to be tested though

### CODE REVIEW

**Comment by omegas27** [comment]
Ingress:
```yaml
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  annotations:
    kubernetes.io/ingress.class: nginx
    kubernetes.io/tls-acme: "true"
    nginx.ingress.kubernetes.io/auth-signin: https://$host/oauth2/start?rd=$escaped_request_uri
    nginx.ingress.kubernetes.io/auth-url: https://$host/oauth2/auth
    nginx.ingress.kubernetes.io/whitelist-source-range: 185.102.95.187/32,95.179.153.65/32
  creationTimestamp: "2020-09-24T13:38:03Z"
  generation: 1
  labels:
    app.kubernetes.io/instance: godsmack
    app.kubernetes.io/managed-by: prometheus-meta-operator
    app.kubernetes.io/name: prometheus
  name: godsmack
  namespace: godsmack-prometheus
  resourceVersion: "205509928"
  selfLink: /apis/extensions/v1beta1/namespaces/godsmack-prometheus/ingresses/godsmack
  uid: 87b27fa8-2419-46a0-a722-f860cc66c66c
spec:
  rules:
  - host: prometheus.g8s.godsmack.westeurope.azure.gigantic.io
    http:
      paths:
      - backend:
          serviceName: godsmack
          servicePort: web
        path: /godsmack
  tls:
  - hosts:
    - prometheus.g8s.godsmack.westeurope.azure.gigantic.io
    secretName: prometheus-certificates
status:
  loadBalancer:
    ingress:
    - {}
```

Service

```yaml
apiVersion: v1
kind: Service
metadata:
  annotations:
    giantswarm.io/monitoring-path: metrics
    giantswarm.io/monitoring-port: web
  creationTimestamp: "2020-09-24T12:57:14Z"
  labels:
    app.kubernetes.io/instance: godsmack
    app.kubernetes.io/managed-by: prometheus-meta-operator
    app.kubernetes.io/name: prometheus
    giantswarm.io/monitoring: "true"
  name: godsmack
  namespace: godsmack-prometheus
  resourceVersion: "205498212"
  selfLink: /api/v1/namespaces/godsmack-prometheus/services/godsmack
  uid: 0b56619f-3d45-409a-bafe-749ba7a15c4c
spec:
  clusterIP: 172.31.0.116
  ports:
  - name: web
    port: 9090
    protocol: TCP
    targetPort: web
  selector:
    app.kubernetes.io/instance: godsmack
    app.kubernetes.io/managed-by: prometheus-meta-operator
    app.kubernetes.io/name: prometheus
  sessionAffinity: None
  type: ClusterIP
status:

```

File: files/templates/additional-scrape-configs.template.yaml
Comment (mig4) [inline files/templates/additional-scrape-configs.template.yaml:2485]: Why delete this?

File: helm/prometheus-meta-operator/values.yaml
Comment (mig4) [inline helm/prometheus-meta-operator/values.yaml:5]: Are you sure this is right?

File: main.go
Comment (mig4) [inline main.go:120]: Would you mind just even describing in the PR why add these flags?

File: service/controller/resource/alert/rules/labellingschema.go
Comment (mig4) [inline service/controller/resource/alert/rules/labellingschema.go:43]: Is this related?

File: service/controller/resource/ingress/resource.go
Comment (mig4) [inline service/controller/resource/ingress/resource.go:68]: Just out of curiosity, when is that false? I.e. when do we not want TLS enabled?

File: service/controller/resource/prometheus/resource.go
Comment (mig4) [inline service/controller/resource/prometheus/resource.go:63]: Is this related?

File: files/templates/additional-scrape-configs.template.yaml
Comment (omegas27) [reply files/templates/additional-scrape-configs.template.yaml:2485]: Because I needed to create a service for the ingress and added the annotations to the service so this is getting picked up by the managed apps part

File: main.go
Comment (omegas27) [reply main.go:120]: I added them to keep the behavior here: https://github.com/giantswarm/g8s-prometheus/blob/master/helm/g8s-prometheus/templates/prometheus/ingress.yaml

Basically, some installations do not support letsencrypt and some restrict access to prometheus to only some subnets

Comment (omegas27) [reply main.go:120]: Where do you think I should add something?

File: service/controller/resource/alert/rules/labellingschema.go
Comment (omegas27) [reply service/controller/resource/alert/rules/labellingschema.go:43]: No, I just discovered the intstr.FromString util function during this PR

File: service/controller/resource/ingress/resource.go
Comment (omegas27) [reply service/controller/resource/ingress/resource.go:68]: Some installations do not support letsencrypt so the customers provide us their own certificates hence this become false

File: service/controller/resource/prometheus/resource.go
Comment (omegas27) [reply service/controller/resource/prometheus/resource.go:63]: This is the same code

File: helm/prometheus-meta-operator/values.yaml
Comment (omegas27) [reply helm/prometheus-meta-operator/values.yaml:5]: Nope, let me fix it

File: files/templates/additional-scrape-configs.template.yaml
Comment (mig4) [reply files/templates/additional-scrape-configs.template.yaml:2485]: Got it, thanks.

File: main.go
Comment (mig4) [reply main.go:120]: Nice, I didn't know about that 👍 

Well I try to describe every change in the PR description so it ends up in the git comment to master. This helps to figure out why something was done when going through history.

File: service/controller/resource/ingress/resource.go
Comment (mig4) [reply service/controller/resource/ingress/resource.go]: Thanks, I didn't know about that.

File: main.go
Comment (omegas27) [reply main.go:120]: I will rewrite the commit before the merge then :D

File: helm/prometheus-meta-operator/templates/configmap.yaml
Comment (mig4) [inline helm/prometheus-meta-operator/templates/configmap.yaml:40]: I know, this is very much a nitpick but no newline at the end of file triggers my OCD 😉 

```suggestion
        host: {{ .Values.Installation.V1.Auth.Vault.Host }}

```

File: service/controller/resource/ingress/resource.go
Comment (mig4) [inline service/controller/resource/ingress/resource.go:96]: Just for my own understanding, I didn't think you needed to specify `TypeMeta` explicitly, what's the reason for it?

File: helm/prometheus-meta-operator/templates/rbac.yaml
Comment (mig4) [inline helm/prometheus-meta-operator/templates/rbac.yaml:73]: Do we need all those?

**Review by mig4 — APPROVED** [review]
Very nice and comprehensive, LGTM when tested + a couple of questions.

File: service/controller/resource/ingress/resource.go
Comment (omegas27) [reply service/controller/resource/ingress/resource.go:96]: I think we do not really need it but we specify it elsewhere. I will do a PR to remove it after we are done with the rest of the story :D

File: service/controller/control-plane/resource.go
Comment (mig4) [inline service/controller/control-plane/resource.go:30]: Is this still needed?

File: helm/prometheus-meta-operator/templates/rbac.yaml
Comment (omegas27) [reply helm/prometheus-meta-operator/templates/rbac.yaml:73]: We probably do not need list, but that should be it. I think we need to refactor some part of the project after this story

File: flag/service/security/security.go
Comment (mig4) [inline flag/service/security/security.go:4]: Is this still needed (given removal of Let's Encrypt)?

**Review by mig4 — APPROVED** [review]

File: flag/service/security/security.go
Comment (omegas27) [reply flag/service/security/security.go:4]: Well, we will need it for oauth2-proxy so I was thinking it was not really useful to remove it

File: main.go
Comment (omegas27) [reply main.go:120]: Done

File: helm/prometheus-meta-operator/templates/rbac.yaml
Comment (TheoBrigitte) [reply helm/prometheus-meta-operator/templates/rbac.yaml:73]: I think we only need 
```
- create
- delete
- get
- update
```

File: service/controller/control-plane/resource.go
Comment (TheoBrigitte) [inline service/controller/control-plane/resource.go:150]: What is this for? There's already a `prometheus-operated` service created by prometheus-operator,

File: service/controller/resource/alert/rules/labellingschema.go
Comment (TheoBrigitte) [inline service/controller/resource/alert/rules/labellingschema.go:43]: Can we move this to a different PR please.

File: helm/prometheus-meta-operator/templates/rbac.yaml
Comment (omegas27) [reply helm/prometheus-meta-operator/templates/rbac.yaml:73]: Lets try

File: service/controller/control-plane/resource.go
Comment (omegas27) [reply service/controller/control-plane/resource.go:30]: See below

File: service/controller/resource/ingress/resource_test.go
Comment (omegas27) [reply service/controller/resource/ingress/resource_test.go:13]: Done

File: service/controller/control-plane/resource.go
Comment (omegas27) [reply service/controller/control-plane/resource.go:150]: This service is for the statefulset to actually work and does not have a cluster IP so we cannot use it for the ingress

File: files/templates/additional-scrape-configs.template.yaml
Comment (omegas27) [reply files/templates/additional-scrape-configs.template.yaml:2486]: Because the service I created for the ingress has the annotations to make it get scraped in the managed apps section

File: service/controller/resource/alert/rules/labellingschema.go
Comment (omegas27) [reply service/controller/resource/alert/rules/labellingschema.go:43]: Sure

Comment (omegas27) [reply service/controller/resource/alert/rules/labellingschema.go:43]: https://github.com/giantswarm/prometheus-meta-operator/pull/142

File: service/controller/control-plane/resource.go
Comment (TheoBrigitte) [reply service/controller/control-plane/resource.go:152]: This works I did it there https://github.com/giantswarm/giantswarm/issues/11524#issuecomment-696832297

Comment (omegas27) [reply service/controller/control-plane/resource.go:152]: Cool, so we can do without, but if we manage it, we can add annotations to it which we cannot on the statefulset service

**Comment by TheoBrigitte** [comment]
Tested successfully on gauss

image

image

**Comment by TheoBrigitte** [comment]
Let's test this on a kvm installation where we have custom certificates also.

**Comment by omegas27** [comment]
Testing on geckon

**Comment by omegas27** [comment]
Both our onprem testing installations are using lets encrypt

**Comment by omegas27** [comment]
That means we need to test on a customer installation. Are we fine with that?

**Review by TheoBrigitte — APPROVED** [review]

File: flag/service/security/security.go
Comment (omegas27) [reply flag/service/security/security.go]: Turns out we do not now so I removed it

**Pull request merged**
